src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -13,47 +33,143 @@
 import io.supertokens.totp.exceptions.LimitReachedException;
 
 public class Totp {
+    private static String generateSecret() throws Exception {
+        final String TOTP_ALGORITHM = "HmacSHA1";
+
+        try {
+            final KeyGenerator keyGenerator = KeyGenerator.getInstance(TOTP_ALGORITHM);
+            keyGenerator.init(160); // 160 bits = 20 bytes
 
-    public static String generateSecret() {
-        return "XXXX";
+            // FIXME: Should return base32 or base16
+            // Return base64 string of the secret key:
+            return Base64.getEncoder().encodeToString(keyGenerator.generateKey().getEncoded());
+        } catch (NoSuchAlgorithmException e) {
+            throw new Exception("TOTP algorithm not found");
+        }
     }
 
-    public static boolean checkCode(TOTPDevice device, String code) {
-        if (code.startsWith("XXXX")) {
-            return true;
+    private static boolean checkCode(TOTPDevice device, String code) {
```

Comment:
where did you get the code for this function from?

Reply:
Wrote on my own. We need to check the codes for different skews.

Reply:
You should find literature about this, or see in other open source libs. Or then clearly explain how you derived the logic for this code.

Reply:
The logic is straightforward but here's a reference: https://github.com/pquerna/otp/blob/87d222c6866be9f43282a4ec79330a76c10f2e18/totp/totp.go#L98

This lib is used by https://github.com/authelia/authelia

Basically, skew decides how many periods should we look before as well as after the current one. This ensures that even if there's a slight time lag between the server and the client app (say Authy), the server will still accept the code. (leading to be a better UX). skew = 1 is the sufficient for most cases.
 
So if skew = 0, we just check (2*0 + 1) = 1 period. This means just the current period
If skew = 1, we check (2*1 + 1) = 3 periods. This means the just previous, current, and immediate next period.
If it's 2, we check (2*2 + 1) = 5 periods. This means the previous 2, the current one, and the next 2.

The logic that I've written simply does that. Iterates over different skews.  `for (int i = -skew; i <= skew; i++)`
and then the period value is passed to the java-otp lib using `Instant.now().plusSeconds(i * period)`.